esp-wifi/src/preempt/mod.rs

```diff
@@ -1,49 +1,102 @@
 
-macro_rules! task_stack {
-    ($($task_stack_size:literal),+) => {
-        const TASK_COUNT: usize = [$($task_stack_size),+].len();
-        const TASK_STACK_SIZE: [usize; TASK_COUNT] = [$($task_stack_size),+];
-        const TOTAL_STACK_SIZE: usize = sum!($($task_stack_size),+);
-        const MAX_TASK: usize = TASK_COUNT + 1; // +1 for the user program
+pub fn allocate_main_task() -> *mut Context {
+    critical_section::with(|_| unsafe {
+        if !CTX_NOW.is_null() {
+            panic!("Tried to allocate main task multiple times");
+        }
 
-        static mut TASK_STACK: [u8; TOTAL_STACK_SIZE] = [0u8; TOTAL_STACK_SIZE];
-    };
+        let ptr = malloc(size_of::<Context>() as u32) as *mut Context;
+        core::ptr::write_bytes(ptr, 0, 1);
+        (*ptr).next = ptr;
+        CTX_NOW = ptr;
+        ptr
+    })
 }
 
-static mut TASK_TOP: usize = 1;
-static mut CTX_NOW: usize = 0;
+fn allocate_task() -> *mut Context {
+    critical_section::with(|_| unsafe {
+        if CTX_NOW.is_null() {
+            panic!("Called `allocate_task` before allocating main task");
+        }
 
-fn allocate_task() -> usize {
-    unsafe {
-        let i = TASK_TOP - 1;
-        CTX_NOW = TASK_TOP;
-        TASK_TOP += 1;
-        i
-    }
+        let ptr = malloc(size_of::<Context>() as u32) as *mut Context;
+        core::ptr::write_bytes(ptr, 0, 1);
+        (*ptr).next = (*CTX_NOW).next;
+        (*CTX_NOW).next = ptr;
+        ptr
+    })
 }
 
 fn next_task() {
     unsafe {
-        CTX_NOW = (CTX_NOW + 1) % TASK_TOP;
+        CTX_NOW = (*CTX_NOW).next;
     }
 }
 
-pub fn current_task() -> usize {
+fn remove_task(task: *mut Context) {
+    critical_section::with(|_| unsafe {
+        let mut ptr = CTX_NOW;
+        loop {
+            if (*ptr).next == task {
+                (*ptr).next = (*(*ptr).next).next;
+
+                free((*task).allocated_stack as *mut u8);
+                free(task as *mut u8);
+                break;
+            }
+
+            ptr = (*ptr).next;
+
+            if ptr == CTX_NOW {
+                break;
+            }
+        }
+
+        memory_fence();
+    });
+}
+
+pub fn current_task() -> *mut Context {
     unsafe { CTX_NOW }
 }
 
-#[cfg(coex)]
-task_stack!(8192, 8192, 8192);
+#[cfg(feature = "wifi")]
+pub fn schedule_task_deletion(task: *mut Context) {
+    use crate::timer::yield_task;
 
-#[cfg(not(coex))]
-task_stack!(8192, 8192);
+    unsafe {
+        SCHEDULED_TASK_TO_DELETE = task;
+    }
 
-#[cfg_attr(target_arch = "riscv32", path = "preempt_riscv.rs")]
-#[cfg_attr(target_arch = "xtensa", path = "preempt_xtensa.rs")]
-#[allow(clippy::module_inception)]
-pub mod preempt;
+    if task == current_task() {
+        loop {
+            yield_task();
+        }
+    }
+}
+
+pub fn task_switch(trap_frame: &mut TrapFrame) {
+    save_task_context(current_task(), trap_frame);
+
+    unsafe {
+        if !SCHEDULED_TASK_TO_DELETE.is_null() {
```

Comment:
Optional: Could we clean this up a bit with `core::ptr::replace`?

Reply:
Not quite sure how - but might be because it feels like 50 degree in my office right now 😓

Reply:
I thought it might be helpful, but I can't get it to work after playing with it. Unfortunately, I can't blame the heat because it's raining :smiling_face_with_tear: